python/task_manager/results_poller.py

```diff
@@ -9,103 +9,179 @@
+    to the background task manager thread in a thread-safe manner, since it
+    doesn't exhibit the bad behaviour from PySide's signals.
+    """
+
+    class _ShutdownHint(object):
+        """
+        Hint to dispatcher that it should shut down.
+        """
+
     # Emitted when a task is completed.
     task_completed = QtCore.Signal(object, object, object)
     # Emitted when a task has failed.
     task_failed = QtCore.Signal(object, object, object, object)
 
-    _POLLING_INTERVAL = 100
-
     def __init__(self, parent=None):
         """
         Constructor.
 
         :param parent:  The parent QObject for this thread
         """
-        QtCore.QObject.__init__(self, parent)
-        # Results queue that will be polled
+        QtCore.QThread.__init__(self, parent)
+        # Results that will need to be dispatched to the background task
+        # manager.
         self._results = Queue.Queue()
-        # Create a timer with no interval. This means the timer will be invoked
-        # only when the event queue is empty.
-        self._timer = QtCore.QTimer(parent=self)
-        self._timer.setInterval(self._POLLING_INTERVAL)
-        # Since the event will be invoked as soon as the event queue is empty,
-        # we should process the queue one item at a time to not interfere with the
-        # ui responsivenes
-        self._timer.timeout.connect(self._flush_events)
         self._bundle = sgtk.platform.current_bundle()
 
-    def start(self):
+    def _log(self, msg):
         """
-        Start polling for tasks that ended.
+        Logs a message at the debug level.
         """
-        self._timer.start()
+        self._bundle.log_debug("Results Queue: %s" % msg)
 
     def shut_down(self):
         """
-        Stop polling for tasks that ended.
+        Shuts down the result dispatcher thread.
         """
-        self._timer.stop()
+        # Add an event in the queue that will tell the thread to terminate.
+        self._results.put(self._ShutdownHint())
+        self._log("Sent _ShutdownHint to consumer thread.")
+        # Do not wait for the thread here!!! The backgroud thread is invoking
+        # the main thread synchronously. Waiting here would introduce
+        # a deadlock because the main thread would be waiting for the dispatcher
+        # to end and the dispatcher would be waiting for his events to
+        # be processed by the main thread.
 
-    def _flush_events(self):
+    def run(self):
         """
         Executes callbacks for each task result.
         """
+        while True:
+            # Wait for the next result
+            self._event = self._results.get(block=True)
+            # If we're told that we need to quit, do it.
+            if isinstance(self._event, self._ShutdownHint):
+                self._log("Consumer thread received ShutdownHint.")
+                break
+
+            # In order to keep this loop simple, we will assume that the background
```

Comment:
Useful wall of text to explain how everything here is thread-safe and simpler this way.